Feature/aws reduce s3 handler mem usage

otherchirps opened the pull request:
### What does this PR do?

Reduce the memory consumption by the aws lambda function, mainly when processing gzipped logs.

### Motivation

When reading `.gz` logs, to send back to datadog, multiple in-memory copies of the decompressed data are being made.  This is ok in most cases, but when the compressed files are already large (eg. 100MB+), the lambda function tended to bomb out with Out of Memory exceptions.

Previously, the `s3_handler` would:

1. Read the entire contents of the gzipped file into memory from s3 (not too much choice here...),
2. Open a `gzip.GzipFile` iterator to walk each decompressed line of the file,
3. Immediately join all the lines ("all lines" list, unneeded copy 1) into one big `data` string (unneeded copy 2),
4. For the case of just "Send lines to Datadog", the `data` string is again split back in to a list of lines (unneeded copy 3).

This change tries to eliminate the copies created in steps 3 & 4.

In our local test case, memory usage against a 100MB gzip logfile dropped from approx 3GB (and crashing...) down to 300MB (and completing...).

For the other cases catered for in this handler (`is_cloudtrail` and has a `DD_MULTILINE_LOG_REGEX_PATTERN`), the behaviour remains the same as before.

For the `is_cloudtrail` case, it expects `data` to be a big blob of json to be parsed.
Whereas the multi-line regex check expects many lines in `data`.  So both of these have been left alone for now.

### Additional Notes

`normalize_logs` and `DatadogBatcher.batch` were consuming the log generators they were given and returning full lists (extra copies of all the log data).  Also altered these to be generator functions.

otherchirps committed:
[aws] Reduce s3_handler memory usage for processing gzipped logs.

When reading `.gz` logs, to send back to datadog, multiple in-memory
copies of the decompressed data are being made.  This is ok in most cases, but when
the compressed files are already large (eg. 100MB+), the lambda function
tends to bomb out with Out of Memory exceptions.

Previously, the `s3_handler` would:

1. Read the entire contents of the gzipped file into memory from s3 (not too much choice here...),
2. Open a `gzip.GzipFile` iterator to walk each decompressed line of the file,
3. Immediately join all the lines ("all lines" list, unneeded copy 1) into one big `data` string (unneeded copy 2),
4. For the case of just "Send lines to Datadog", the `data` string is again split back in to a list of lines (unneeded copy 3).

This change tries to eliminate the copies created in steps 3 & 4.

In our local test case, memory usage against a 100MB gzip logfile dropped
from approx 3GB (and crashing...) down to 300MB (and completing...).

For the other cases catered for in this handler (`is_cloudtrail` and has a `DD_MULTILINE_LOG_REGEX_PATTERN`),
the behaviour remains the same as before.

For the `is_cloudtrail` case, it expects `data` to be a big blob of json to be parsed.
Whereas the multi-line regex check expects many lines in `data`...  May need to
rethink the approach for those if they prove problematic as well.

otherchirps committed:
[aws] Convert normalize_logs into generator.

`generate_logs` is called on by the batcher, which expects an iterable
to be returned from `generate_logs`.  From the name, I assume that it
was intended to be a generator, rather than returning an entire list at
once.

`normalize_logs` was being used, at the end of `generate_logs` to mix
some metadata into each log entry.

Unfortunately, `normalize_logs` was returning a regular list, rather
than the generator yielded by the event handlers (eg. s3_handler,
awslogs_handler).  Instead of yielding each entry, it consumed the log
generator it had been given, building and returning the entire
list.

That's fine for small volumes of logs, but letting things remain
generators has a much better memory footprint.

This change alters `normalize_logs` to yield each entry, after attaching
the desired metadata.

otherchirps committed:
[aws] Convert DatadogBatcher.batch into a generator.

Similar to the previous few changes, the batch function was consuming
its given generator entirely before returning the full result list.

This change converts the routine into a generator, so it only yields a
batch at a time, instead of all the batches of all the logs.  This
prevents another potentially large list being created.

otherchirps commented:
👋 Hi folks, just checking if anyone has any problems or questions with these changes?  
Understand everyone is likely busy, so mightn't have had a chance to look yet.  But putting it out there that I'm happy to talk through the changes / adjust the PR if necessary.

fwiw, have been using these changes in production here at my workplace for a while now, and has seen a significant reduction in memory usage by the lambda function.  I didn't attempt to make changes to the azure routine, since we're an aws-only shop at the moment.

NilsBunge commented:
Hello @otherchirps ,

Thanks a lot for submitting this PR with all this context and details. 
I'll make sure to forward this to team so they can review it and make any comment if any.

otherchirps commented:
Thanks @NBParis. I notice that there's now a conflict with master though.  I'll try to submit an update to resolve that in the next day or two.

otherchirps committed:
[aws] Convert filter_logs into a generator

otherchirps committed:
[aws] Removed early return from filter_logs.

Previously converted filter_logs into a generator function, to avoid
creating extra copies of the `logs` list.

However, since support for python2.x is expected, `return <value>` is illegal
inside a generator function.

I believe this check was to short-circuit the for loop, in
`filter_logs`, and skip checking for `EXCLUDE_AT_MATCH` and
`INCLUDE_AT_MATCH` for each log.  But I don't see the checking of these
variables against `None` to be costly.  The regex searches if they're
not `None` is the expensive part. So removing the short-circuit before
the for loop, in my mind, is ok.

Another option could be to replace it with a for-yield loop as well, but
I felt that might just be confusing.

otherchirps commented:
> I notice that there's now a conflict with master though. I'll try to submit an update to resolve that in the next day or two.

Well, that was a long "day or two". My apologies.

I believe this change is ready again, and it's working for us in our environments.  Let me know if anything needs clarifying/changing.

ajacquemot commented on aws/logs_monitoring/lambda_function.py:
```diff
@@ -591,7 +584,23 @@ def s3_handler(event, context, metadata):
+
+                # NOTE: Duplicates the non-compressed case below, but we can't allow falling through to that
+                #   since we'll lose the decompress_stream context.
+                for line in decompress_stream:
```
this block is duplicate from https://github.com/DataDog/datadog-serverless-functions/pull/117/files#diff-618731ddf1c446ef45d09613cb4b189bR622, do you think we can factorise this?

otherchirps replied:
Good call.  Yes, I'll push through an update shortly.

ajacquemot reviewed (approved):
Hi @otherchirps,

Thanks a lot for working on this, I left one comment but I don't think this should be a blocker.
Cheers

otherchirps committed:
[aws] Centralise s3_handler event creation

ajacquemot commented on aws/logs_monitoring/lambda_function.py:
```diff
@@ -572,6 +565,25 @@ def s3_handler(event, context, metadata):
+    }
+
+    # Build a structured message out of the s3 event data.
+    def reformat_record(record):
```
nits
```suggestion
    def build_message(record):
```
or
```suggestion
    def to_message(record):
```

otherchirps replied:
No problem.  Renamed to `build_message`.

otherchirps committed:
[aws] Rename nested function: reformat_record -> build_message

otherchirps commented:
Rebased on current master to resolve merge conflict.

tianchu commented:
Opened a new PR with @otherchirps commits rebased against the latest master https://github.com/DataDog/datadog-serverless-functions/pull/210

tianchu commented:
Please try out the new approach to ingest logs from AWS to Datadog using the [Datadog Kinesis firehose destination](https://docs.datadoghq.com/logs/guide/send-aws-services-logs-with-the-datadog-kinesis-firehose-destination/).
